### ISSUE TITLE
deckjs backend failing on some machines

### ISSUE THREAD

**Issue by tlberglund** [asciidoctor/asciidoctor-gradle-plugin#153]
I put some of these comments in #106, but since that issue is closed, I figured I'd open a new one. We are seeing that behavior on two machines on our team, and not on another. 

I put the [build file in a Gist](https://gist.github.com/tlberglund/5317eda1bc2a80dd9145). When we run `gradle presentation --stacktrace`, we get the following:

```
Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':presentation'.
at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:69)
at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:46)
at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:35)
at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:64)
at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:42)
at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
at org.gradle.api.internal.AbstractTask.executeWithoutThrowingTaskFailure(AbstractTask.java:305)
at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.executeTask(AbstractTaskPlanExecutor.java:79)
at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:63)
at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:51)
at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(DefaultTaskPlanExecutor.java:23)
at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:88)
at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:29)
at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:62)
at org.gradle.execution.DefaultBuildExecuter.access$200(DefaultBuildExecuter.java:23)
at org.gradle.execution.DefaultBuildExecuter$2.proceed(DefaultBuildExecuter.java:68)
at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:62)
at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:55)
at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:149)
at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:106)
at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:86)
at org.gradle.launcher.exec.InProcessBuildActionExecuter$DefaultBuildController.run(InProcessBuildActionExecuter.java:80)
at org.gradle.launcher.cli.ExecuteBuildAction.run(ExecuteBuildAction.java:33)
at org.gradle.launcher.cli.ExecuteBuildAction.run(ExecuteBuildAction.java:24)
at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:36)
at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:26)
at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:51)
at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:171)
at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:237)
at org.gradle.launcher.cli.CommandLineActionFactory$ParseAndBuildAction.execute(CommandLineActionFactory.java:210)
at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:35)
at org.gradle.launcher.cli.JavaRuntimeValidationAction.execute(JavaRuntimeValidationAction.java:24)
at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:206)
at org.gradle.launcher.cli.CommandLineActionFactory$WithLogging.execute(CommandLineActionFactory.java:169)
at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:33)
at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:22)
at org.gradle.launcher.Main.doAction(Main.java:33)
at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:45)
at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:54)
at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:35)
at org.gradle.launcher.GradleMain.main(GradleMain.java:23)
at org.gradle.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:30)
at org.gradle.wrapper.WrapperExecutor.execute(WrapperExecutor.java:127)
at org.gradle.wrapper.GradleWrapperMain.main(GradleWrapperMain.java:56)
Caused by: org.gradle.api.GradleException: Error running Asciidoctor
at org.asciidoctor.gradle.AsciidoctorTask.processDocumentsAndResources(AsciidoctorTask.groovy:659)
at org.asciidoctor.gradle.AsciidoctorTask.this$4$processDocumentsAndResources(AsciidoctorTask.groovy)
at org.asciidoctor.gradle.AsciidoctorTask$this$4$processDocumentsAndResources.callCurrent(Unknown Source)
at org.asciidoctor.gradle.AsciidoctorTask.processAsciidocSources(AsciidoctorTask.groovy:592)
at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:63)
at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.doExecute(AnnotationProcessingTaskFactory.java:218)
at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:211)
at org.gradle.api.internal.project.taskfactory.AnnotationProcessingTaskFactory$StandardTaskAction.execute(AnnotationProcessingTaskFactory.java:200)
at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:579)
at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:562)
at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:80)
at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:61)
... 47 more
Caused by: org.jruby.exceptions.RaiseException: (RuntimeError) Could not find a converter to handle transform: document
at RUBY.find_converter(/Users/Cjmiller-RMP15/.gradle/caches/modules-2/files-2.1/org.asciidoctor/asciidoctorj/1.5.2/39d33f739ec1c46f6e908a725264eb74b23c9f99/asciidoctorj-1.5.2.jar!/gems/asciidoctor-1.5.2/lib/asciidoctor/converter/composite.rb:52)
at RUBY.converter_for(/Users/Cjmiller-RMP15/.gradle/caches/modules-2/files-2.1/org.asciidoctor/asciidoctorj/1.5.2/39d33f739ec1c46f6e908a725264eb74b23c9f99/asciidoctorj-1.5.2.jar!/gems/asciidoctor-1.5.2/lib/asciidoctor/converter/composite.rb:41)
at RUBY.convert(/Users/Cjmiller-RMP15/.gradle/caches/modules-2/files-2.1/org.asciidoctor/asciidoctorj/1.5.2/39d33f739ec1c46f6e908a725264eb74b23c9f99/asciidoctorj-1.5.2.jar!/gems/asciidoctor-1.5.2/lib/asciidoctor/converter/composite.rb:31)
at RUBY.convert(/Users/Cjmiller-RMP15/.gradle/caches/modules-2/files-2.1/org.asciidoctor/asciidoctorj/1.5.2/39d33f739ec1c46f6e908a725264eb74b23c9f99/asciidoctorj-1.5.2.jar!/gems/asciidoctor-1.5.2/lib/asciidoctor/document.rb:1028)
at RUBY.convert(/Users/Cjmiller-RMP15/.gradle/caches/modules-2/files-2.1/org.asciidoctor/asciidoctorj/1.5.2/39d33f739ec1c46f6e908a725264eb74b23c9f99/asciidoctorj-1.5.2.jar!/gems/asciidoctor-1.5.2/lib/asciidoctor.rb:1499)
at RUBY.convert_file(/Users/Cjmiller-RMP15/.gradle/caches/modules-2/files-2.1/org.asciidoctor/asciidoctorj/1.5.2/39d33f739ec1c46f6e908a725264eb74b23c9f99/asciidoctorj-1.5.2.jar!/gems/asciidoctor-1.5.2/lib/asciidoctor.rb:1562)
at RUBY.convertFile(<script>:68)
at org.jruby.gen.InterfaceImpl748868100.convertFile(org/jruby/gen/InterfaceImpl748868100.gen:13)
at org.asciidoctor.gradle.AsciidoctorProxyImpl.renderFile(AsciidoctorProxyImpl.groovy:26)
at org.asciidoctor.gradle.AsciidoctorTask.processSingleFile(AsciidoctorTask.groovy:672)
at org.asciidoctor.gradle.AsciidoctorTask$_processDocumentsAndResources_closure7.doCall(AsciidoctorTask.groovy:652)
at org.asciidoctor.gradle.AsciidoctorTask.processDocumentsAndResources(AsciidoctorTask.groovy:647)
at org.asciidoctor.gradle.AsciidoctorTask.processAsciidocSources(AsciidoctorTask.groovy:592)
```

I've been doing as much digging into this as I can myself, but any help would be appreciated. :grinning:

**Comment by tlberglund** [comment]
I ran the build with `--debug`, and extracted the literal command line string the plugin was generating:

```
asciidoctor -B /Users/Cjmiller-RMP15/dev/ds-training/instructional-techniques/src/asciidoc -T /Users/Cjmiller-RMP15/dev/ds-training/instructional-techniques/src/framework/asciidoctor-backends/haml -e erubis -S UNSAFE -b deckjs -a source-highlighter=coderay -a idprefix -a idseparator=- -a projectdir=../.. -a rootdir=../.. -a project-name=instructional-techniques -a project-group -a project-version=unspecified /Users/Cjmiller-RMP15/dev/ds-training/instructional-techniques/src/asciidoc/slides.adoc
```

When I run that string at the terminal, I get the help output. Some trial and error revealed that changing `-S UNSAFE` to `-S unsafe` averted this problem, and got us back to the `org.jruby.exceptions.RaiseException: (RuntimeError) Could not find a converter to handle transform: document` exception. That exception happens on the machines on which the build breaks, and does not happen on the machine on which the build runs.

I don't know if this debug string is intended to be interpreted the way I am interpreting it, but I thought this was interesting and worth sharing.

**Comment by aalmiray** [comment]
I understand this as sending "UNSAFE" causes some trouble. However changing it to "unsafe" does not fix the problem altogether, right?

**Comment by tlberglund** [comment]
Correct. I just threw that in as an extra clue in case it was helpful. On that point, is there an option in the plugin config that changes how that parameter is set when asciidoctor runs?

**Comment by aalmiray** [comment]
I may be wrong but I think Asciidoctorj is the one setting the value in upper case as `AsciidoctorTask` attempts a resolution from int/CharSequence/enum into int, as seen at https://github.com/asciidoctor/asciidoctor-gradle-plugin/blob/development/src/main/groovy/org/asciidoctor/gradle/AsciidoctorTask.groovy#L820-L834

You can try using different int values and/or lowercase names from https://github.com/asciidoctor/asciidoctorj/blob/master/asciidoctorj-core/src/main/java/org/asciidoctor/SafeMode.java

@lordofthejars may know more about this behavior.

**Comment by tlberglund** [comment]
So if I had cloned `asciidoctorj` and deployed a Maven-local snapshot with a tweaked-up `SafeMode` class, what dependency would I change in the `asciidoctor-gradle-plugin` build to point it to the snapshot? There's a `testCompile` dependency on `org.asciidoctor:asciidoctorj:1.5.3`, but it doesn't strike me that that will get the job done.

**Comment by ysb33r** [comment]
In the `build.gradle` just set the version you want

```groovy
asciidoctorj {
  version = '1.2.3-SNAPSHOT'
}
```

**Comment by tlberglund** [comment]
Which project's `build.gradle` gets that?

**Comment by lordofthejars** [comment]
Tomorrow morning I will take a look :)
El dt., 10 de març, 2015 a les 20.45 Tim Berglund <notifications@github.com>
va escriure:

**Comment by tlberglund** [comment]
:+1:

**Comment by tlberglund** [comment]
We figured this out. The problem was that the things under the `frameworkDir` path (see linked build file up above, lines 15-18) were empty. I used Git submodules for those components, and forgot to have the two new users `--recurse-submodules` on them. With the submodules initialized, everything is peachy.

The best I can make of this issue is the suggestion that the error message might be better, but the plugin definitely can't be expected to have produced output under these circumstances!

**Comment by aalmiray** [comment]
re: asciidoctorj block. We'll have a look at that as I believe we fixed a related issue in 1.5.3-SNAPSHOT.

:tada: on finding the cause of the problem and figuring it out

**Comment by ysb33r** [comment]
@tlberglund That has to go into the `build.gradle` of your actual documentation project. The feature was added arounf `1.5.0` to allow people the flexibility to use `asciidoctorj` versions other than the default. Therefore if you build your custom `asciidoctorj` JAR, you don't have to rebuild `asciidoctor-gradle-plugin` as well - just tell the consuming project which which version it need to use. (This is described in this plugin's README).
